### PR TITLE
WIP: Split task without generics

### PR DESCRIPTION
**Pull request by justinsb**

### CODE REVIEW

**Comment by justinsb** [comment]
I think the interesting stuff here is all in the `fi` package, the rest is boilerplate.

File: upup/pkg/fi/executor.go [hunk 1]
```diff
@@ -184,7 +184,12 @@ func (e *executor) forkJoin(tasks []*taskState) []error {
 			results[index] = fmt.Errorf("function panic")
 			defer wg.Done()
 			klog.V(2).Infof("Executing task %q: %v\n", ts.key, ts.task)
-			results[index] = ts.task.Run(e.context)
+			if nodeTask, ok := ts.task.(NodeTask); ok {
+				results[index] = nodeTask.Run(e.context.(*NodeContext))
+			} else {
+				cloudTask := ts.task.(CloudTask)
+				results[index] = cloudTask.Run(e.context.(*CloudContext))
+			}
```
Comment (johngmyers) [inline upup/pkg/fi/executor.go:192]: More runtime errors which would be compile time with generics.

File: upup/pkg/fi/task.go [hunk 2]
```diff
@@ -25,13 +25,23 @@ import (
 )
 
 type Task interface {
-	Run(*Context) error
+	HasName
```
Comment (johngmyers) [inline upup/pkg/fi/task.go:28]: I'd like to get it so node tasks don't have to implement unnecessary boilerplate methods such as `GetName`.

File: upup/pkg/fi/has_address.go [hunk 3]
```diff
@@ -19,9 +19,9 @@ package fi
 // HasAddress is implemented by elastic/floating IP addresses in order to include
 // relevant dynamically allocated addresses in the api-server's server TLS certificate.
 type HasAddress interface {
-	Task
+	//Task
```
Comment (johngmyers) [inline upup/pkg/fi/has_address.go:22]: `CloudTask`? I don't want anything not a `Task` to implement this, as such things won't have `FindAddresses()` called.

**Comment by justinsb** [comment]
Addressed by alternate approach

/close

**Pull request closed without merging**